src/sysupdate/sysupdate-transfer.c

```diff
@@ -784,30 +788,193 @@ static void compile_pattern_fields(
+        PidRef pid;
+        void* userdata;
+};
+
+static int helper_on_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        char buf[NOTIFY_BUFFER_MAX+1];
+        struct iovec iovec = {
+                .iov_base = buf,
+                .iov_len = sizeof(buf)-1,
+        };
+        CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred))) control;
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_control = &control,
+                .msg_controllen = sizeof(control),
+        };
+        struct ucred *ucred;
+        struct notify_userdata *ctx = ASSERT_PTR(userdata);
+        char* progress_str;
+        unsigned progress;
+        ssize_t n;
+        int r;
+
+        n = recvmsg_safe(fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
+        if (n < 0) {
+                if (ERRNO_IS_TRANSIENT(n))
+                        return 0;
+                return (int) n;
+        }
+
+        cmsg_close_all(&msghdr);
+
+        if (msghdr.msg_flags & MSG_TRUNC) {
+                log_warning("Got overly long notification datagram, ignoring.");
+                return 0;
+        }
+
+        ucred = CMSG_FIND_DATA(&msghdr, SOL_SOCKET, SCM_CREDENTIALS, struct ucred);
+        if (!ucred || ucred->pid <= 0) {
+                log_warning("Got notification datagram lacking credential information, ignoring.");
+                return 0;
+        }
+        if (ucred->pid != ctx->pid.pid) {
+                log_warning("Got notification datagram from unexpected peer, ignoring.");
+                return 0;
+        }
+
+        buf[n] = 0;
+
+        progress_str = find_line_startswith(buf, "X_IMPORT_PROGRESS=");
+        if (!progress_str)
+                return 0;
+
+        truncate_nl(progress_str);
+        delete_trailing_chars(progress_str, "%");
+
+        r = safe_atou(progress_str, &progress);
+        if (r < 0 || progress > 100) {
+                log_warning("Got invalid percent value '%s', ignoring.", progress_str);
+                return 0;
+        }
+
+        return ctx->callback(ctx->transfer, ctx->instance, progress, ctx->userdata);
+}
+
 static int run_callout(
                 const char *name,
-                char *cmdline[]) {
-
+                char *cmdline[],
+                struct notify_userdata *userdata) {
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_(sd_event_source_unrefp) sd_event_source *exit_source = NULL, *notify_source = NULL;
+        _cleanup_close_ int fd = -EBADF;
+        _cleanup_free_ char *bind_name = NULL;
+        _cleanup_(pidref_done) PidRef pid = PIDREF_NULL;
+        union sockaddr_union bsa;
         int r;
 
         assert(name);
         assert(cmdline);
         assert(cmdline[0]);
 
-        r = safe_fork(name, FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT, NULL);
+        r = sd_event_new(&event);
+        if (r < 0)
+                return r;
```

Comment:
In systemd we generally want that every function either logs about all errors on its own, or about none (in which case it propagates an error code to the caller, so that the caller can log instead). [Note: by "logging" I mean non-debug logging, i.e. LOG_INFO and above, debug level logging (i.e. LOG_DEBUG) is always fine and not considered logging in this context]. 

Now, for each function decide whether it should log on its own, or not. And then stick to it. Typically more "library-like" functions should not log on their own but leave that to the caller, and more "main-function-like" functions should log on their own, but there's a huge grey area in between

Now, run_callout() previously was of the logging kind, i.e the FORK_LOG flag passed to safe_fork() made sure that safe_fork() would log, and there was not too much else, and what there was that had log messages.

Now, you turned this into a function where almost nothing logs anymore except for the pidref_safe_fork() call in the middle (which still retains the FORK_LOG flag). 

Hence, please clean this up. i.e. you probably want to add logging to every call here you make. log_error_errno() typically, since these errors are fatal to the function.